### PR TITLE
Fix pointer invalidation bug in TrimShape.

### PR DESCRIPTION
**Pull request by mookerji**
- ASAN reported a container overflow condition in TrimShape, which
  passes a mutable container reference (shape) and a const reference
  invalidated (end_vertex) by mutations to shape via
  TripLegBuilder::Build. Issue resolved by passing arguments by value.
- Reorders TrimShape arguments to place output reference last.

Fixes some memory access issues reported in:
- https://github.com/valhalla/valhalla/issues/2201

/cc @yuzheyan @kevinkreiser

### CODE REVIEW

**Comment by mookerji** [comment]
Don't worry, will follow up with unit tests!

File: src/thor/triplegbuilder.cc [hunk 1]
```diff
@@ -30,9 +30,9 @@ using namespace valhalla::thor;
 
 namespace {
 
-void TrimShape(const float start,
+void TrimShape(float start,
                PointLL start_vertex,
-               const float end,
+               float end,
```
Comment (mandeepsandhu) [inline src/thor/triplegbuilder.cc]: woohoo! 👍

**Comment by mandeepsandhu** [comment]
> * ASAN reported a container overflow condition in TrimShape, which
>   passes a mutable container reference (shape) and a const reference
>   invalidated (end_vertex) by mutations to shape via
>   TripLegBuilder::Build. Issue resolved by passing arguments by value.

I was looking at `TripLegBuilder::Build`. `end_vrt` is a local var, constructed before we get the `shape` via `tile->edgeinfo(edge->edgeinfo_offset()).shape()`. This local is passed as `const` ref to `TrimShape`. Is this bug caused by `shape` altering this local variable somehow? I guess I don't understand how mutation of shape is altering this local variable which was passed by ref.

**Comment by kevinkreiser** [comment]
@mandeepsandhu see lines 1650 and 1653, those ones were a reference to shape points on the shape to be modified. of course `front` and `back` return references so yeah oops :(

**Comment by mandeepsandhu** [comment]
> @mandeepsandhu see lines 1650 and 1653, those ones were a reference to shape points on the shape to be modified. of course `front` and `back` return references so yeah oops :(

Ah, yes! I was looking at earlier calls to `TrimShape` which was using the locals and didn't see the subsequent ones which are passing refs from the shape object. Thanks, this makers total sense.

File: src/thor/triplegbuilder.cc (hunk 1 above)
Comment (yuzheyan) [reply src/thor/triplegbuilder.cc]: This function also has potential failure when shape is empty. Maybe we need to defend against that. while(current != shape.end() - 1) this is going to blow off when shape is empty

Comment (mookerji) [reply src/thor/triplegbuilder.cc]: Fixed

**Comment by mookerji** [comment]
@yuzheyan @kevinkreiser this is ready to go now.

**Comment by mookerji** [comment]
@kevinkreiser updated to move into the geometry header.

**Comment by dnesbitt61** [comment]
There is a method named trim_polyline in midgard/util. Is TrimShape duplicating this functionality (apart from replacing the vertices with the ones provided)?

File: test/util_midgard.cc [hunk 2]
```diff
@@ -378,6 +378,34 @@ TEST(UtilMidgard, TestSimilarAndEqual) {
...
+  float start = 42.2698097;
+  float end = 47;
+  PointLL start_vertex{8.54709148, 47.3651924};
+  PointLL end_vertex{8.54711914, 47.3651543};
+  std::vector<PointLL> shape = {
+      PointLL{8.5468483, 47.3655319},
+      PointLL{8.54691314, 47.365448},
+      PointLL{8.54711914, 47.3651543},
+  };
+  TrimShape(start, start_vertex, end, end_vertex, shape);
```
Comment (kevinkreiser) [inline test/util_midgard.cc]: woudlnt this illustrate the ASAN fix better if you used `shape.front()` and `shape.back()`?

File: src/midgard/util.cc [hunk 3]
```diff
@@ -89,6 +89,45 @@ template <class container_t> container_t trim_front(container_t& pts, const floa
   return result;
 }
 
+void TrimShape(float start,
```
Comment (mandeepsandhu) [inline src/midgard/util.cc]: minor nit: camel-cased `TrimShape` looks inconsistent with the rest of the convention used in this file, which is snake-case. Maybe rename this to `trim_shape`?

File: test/util_midgard.cc (hunk 2 above)
Comment (mookerji) [reply test/util_midgard.cc]: Fixed

File: src/midgard/util.cc (hunk 3 above)
Comment (mookerji) [reply src/midgard/util.cc]: Fixed, although I'll note that we don't have consistent naming conventions within the larger codebase.

**Comment by mookerji** [comment]
> There is a method named trim_polyline in midgard/util. Is TrimShape duplicating this functionality (apart from replacing the vertices with the ones provided)?

@dnesbitt61 I agree this may be the case, but I'm hoping to fix the problem before embarking on a refactor of the calling function (in a follow up PR). That refactor's going to be tricky, since the calling code is both essential and also lacks any tests.

**Pull request merged**